File: 2025/day_24.ts
type TreeNode = {
  value: string
  left?: TreeNode
  right?: TreeNode
}

function isTreesSynchronized(
  tree1?: TreeNode,
  tree2?: TreeNode
): [boolean, string] {
  function areMirror(a?: TreeNode, b?: TreeNode): boolean {
    if (!a && !b) {
      return true
    }

    if (!a || !b) {
      return false
    }

    if (a.value !== b.value) {
      return false
    }

    return areMirror(a.left, b.right) && areMirror(a.right, b.left)
  }

  const syncronized = areMirror(tree1, tree2)
  return [syncronized, tree1.value]
}

File: 2024/day_24.py
def is_trees_synchronized(tree1, tree2):
  def are_mirror(a, b):
    if not a and not b:
      return True

    if not a or not b:
      return False

    if a["value"] != b["value"]:
      return False

    return (
      are_mirror(a["left"], b["right"]) and 
      are_mirror(a["right"], b["left"])
    )

  syncronized = are_mirror(tree1, tree2)

  return [syncronized, tree1["value"]]
